ddtrace/internal/datadog/profiling/dd_wrapper/src/sample.cpp

```diff
@@ -0,0 +1,336 @@
+#include "sample.hpp"
+#include "global_cache.hpp"
+
+#include <thread>
+
+using namespace Datadog;
+
+Sample::Sample(SampleType type, unsigned int _max_nframes)
+{
+
+    profile_state.one_time_init(type, _max_nframes);
+
+    // Initialize values
+    values.resize(profile_state.get_sample_type_length());
+    std::fill(values.begin(), values.end(), 0);
+    max_nframes = profile_state.get_max_nframes();
+    type_mask = profile_state.get_type_mask();
+
+    // Initialize other state
+    profile_seq = profile_state.get_profile_seq();
+    locations.reserve(max_nframes + 1); // +1 for a "truncated frames" virtual frame
+    cur_label = 0;
+}
+
+ddog_prof_Profile&
+Sample::get_ddog_profile()
+{
+    return profile_state.get_current_profile();
+}
+
+Sample&
+Sample::start_sample()
+{
+    profile_state.entrypoint_check();
+
+    // This initializes the sampling state of the profile.  In particular, it
+    // acquires a collection object from global storage on behalf of the user,
+    // creating it if it isn't already present.
+    auto thread_id = std::this_thread::get_id();
+    return GlobalCache::get(thread_id);
+}
+
+void
+Sample::reset_profile()
+{
+    profile_state.reset();
+}
+
+void
+Sample::push_frame_impl(std::string_view name, std::string_view filename, uint64_t address, int64_t line)
+{
+    static const ddog_prof_Mapping null_mapping = { 0, 0, 0, to_slice(""), to_slice("") };
+    name = profile_state.insert_or_get(name);
+    filename = profile_state.insert_or_get(filename);
+
+    ddog_prof_Location loc = {
+        .mapping = null_mapping, // No support for mappings in Python
+        .function = {
+          .name = to_slice(name),
+          .system_name = {}, // No support for system_name in Python
+          .filename = to_slice(filename),
+          .start_line = 0, // We don't know the start_line for the function
+        },
+        .address = address,
+        .line = line,
+    };
+
+    locations.emplace_back(loc);
+}
+
+void
+Sample::push_frame(std::string_view name, std::string_view filename, uint64_t address, int64_t line)
+{
+
+    if (locations.size() <= max_nframes)
```

Comment:
otherwise we just drop it?

Reply:
Good catch!

To the first degree, yes.

To the second degree, yes, we should also _not_ stash the function/filename stuff, but we absolutely should keep track of how many frames we dropped, since we need to create a virtual frame with that information.